File: src/Services/KeyDetector.cs
using AudioAnalyzer.Interfaces;

namespace AudioAnalyzer.Services;

public class KeyDetector : IKeyDetectorService
{
    private static readonly double[] MajorProfile = { 6.35, 2.23, 3.48, 2.33, 4.38, 4.09, 2.52, 5.19, 2.39, 3.66, 2.29, 2.88 };
    private static readonly double[] MinorProfile = { 6.33, 2.68, 3.52, 5.38, 2.60, 3.53, 2.54, 4.75, 3.98, 2.69, 3.34, 3.17 };

    private static readonly string[] NoteNames = { "C", "C#", "D", "D#", "E", "F", "F#", "G", "G#", "A", "A#", "B" };

    public async Task<(string Key, string Mode, double Confidence)> DetectKeyAsync(string filePath, IProgress<int>? progress = null)
    {
        return await Task.Run(() => DetectKey(filePath, progress));
    }

    public async Task<(string Key, string Mode, double Confidence)> DetectKeyAsync(float[] monoSamples, int sampleRate, IProgress<int>? progress = null)
    {
        return await Task.Run(() => DetectKeyFromSamples(monoSamples, sampleRate, progress));
    }

    public (string Key, string Mode, double Confidence) DetectKey(string filePath, IProgress<int>? progress = null)
    {
        var (monoSamples, sampleRate) = new AudioDataProvider().LoadMono(filePath);
        return DetectKeyFromSamples(monoSamples, sampleRate, progress);
    }

    /// <summary>
    /// Core key detection logic operating on pre-loaded mono samples.
    /// No file I/O occurs in this method.
    /// </summary>
    private (string Key, string Mode, double Confidence) DetectKeyFromSamples(float[] monoSamples, int sampleRate, IProgress<int>? progress = null)
    {
        const int MaxAnalysisSeconds = 30;

        try
        {
            if (monoSamples.Length < sampleRate)
                return ("Unknown", "Unknown", 0);

            // Limit to MaxAnalysisSeconds from center of audio for key detection
            int maxSamples = MaxAnalysisSeconds * sampleRate;
            float[] analysisData;
            if (monoSamples.Length > maxSamples)
            {
                int startOffset = (monoSamples.Length - maxSamples) / 2;
                analysisData = monoSamples.AsSpan(startOffset, maxSamples).ToArray();
                LoggerService.Log($"KeyDetector.DetectKeyFromSamples - Trimmed {monoSamples.Length} -> {maxSamples} samples (center segment)");
            }
            else
            {
                analysisData = monoSamples;
            }

            progress?.Report(20);

            var pcp = ComputePitchClassProfile(analysisData, sampleRate);
            
            progress?.Report(50);

            var (keyIndex, mode, correlation) = FindBestKey(pcp);
            
            progress?.Report(100);

            return (NoteNames[keyIndex], mode == 0 ? "Major" : "Minor", correlation);
        }
        catch (Exception ex)
        {
            LoggerService.Log($"KeyDetector.DetectKeyFromSamples - Error: {ex.Message}");
            return ("Error", "Error", 0);
        }
    }

    private double[] ComputePitchClassProfile(float[] samples, int sampleRate)
    {
        const int fftSize = DspConstants.FFT_SIZE_KEY_DETECTION;
        const int hopSize = 8192;
        const int numBins = 12;
        const double a4Freq = 440.0;

        var pcp = new double[numBins];
        int numFrames = (samples.Length - fftSize) / hopSize + 1;
        if (numFrames <= 0) numFrames = 1;

        var magnitudes = new double[fftSize / 2];

        for (int frame = 0; frame < numFrames; frame++)
        {
            var frameStart = frame * hopSize;
            if (frameStart + fftSize > samples.Length) break;

            Array.Clear(magnitudes, 0, magnitudes.Length);

            var window = samples.AsSpan(frameStart, fftSize);
            ComputeFFTMagnitudes(window, magnitudes, sampleRate);

            for (int pitchClass = 0; pitchClass < numBins; pitchClass++)
            {
                double pitchEnergy = 0;
                double c0Freq = a4Freq * Math.Pow(2, (pitchClass - 9) / 12.0);

                for (int harmonic = 1; harmonic <= 8; harmonic++)
                {
                    double harmonicFreq = c0Freq * harmonic;
                    if (harmonicFreq > sampleRate / 2) break;

                    int freqBin = (int)(harmonicFreq * fftSize / sampleRate);
                    if (freqBin > 0 && freqBin < magnitudes.Length - 1)
                    {
                        double interpolatedMag = magnitudes[freqBin] + 0.5 * (magnitudes[freqBin - 1] + magnitudes[freqBin + 1]);
                        pitchEnergy += interpolatedMag / harmonic;
                    }
                }
                pcp[pitchClass] += pitchEnergy;
            }
        }

        double totalEnergy = pcp.Sum();
        if (totalEnergy > 0)
        {
            for (int i = 0; i < numBins; i++)
                pcp[i] /= totalEnergy;
        }

        return pcp;
    }

    private void ComputeFFTMagnitudes(Span<float> window, double[] magnitudes, int sampleRate)
    {
        int n = magnitudes.Length * 2;
        var complex = new System.Numerics.Complex[n];

        for (int i = 0; i < n; i++)
        {
            double hann = 0.5 * (1 - Math.Cos(2 * Math.PI * i / (n - 1)));
            complex[i] = new System.Numerics.Complex(i < window.Length ? window[i] * hann : 0, 0);
        }

        FFT(complex);

        for (int i = 0; i < magnitudes.Length; i++)
        {
            magnitudes[i] = Math.Sqrt(complex[i].Real * complex[i].Real + complex[i].Imaginary * complex[i].Imaginary);
        }
    }

    private void FFT(System.Numerics.Complex[] data) => FftHelper.FFT(data);

    private (int keyIndex, int mode, double correlation) FindBestKey(double[] pcp)
    {
        double bestCorrelation = -1;
        int bestKey = 0;
        int bestMode = 0;

        for (int root = 0; root < 12; root++)
        {
            var rotatedPcp = RotateArray(pcp, root);

            double majorCorr = ComputeCorrelation(rotatedPcp, MajorProfile);
            if (majorCorr > bestCorrelation)
            {
                bestCorrelation = majorCorr;
                bestKey = root;
                bestMode = 0;
            }

            double minorCorr = ComputeCorrelation(rotatedPcp, MinorProfile);
            if (minorCorr > bestCorrelation)
            {
                bestCorrelation = minorCorr;
                bestKey = root;
                bestMode = 1;
            }
        }

        double normalizedConfidence = Math.Min(1.0, Math.Max(0, (bestCorrelation + 1) / 2));
        return (bestKey, bestMode, normalizedConfidence);
    }

    private double[] RotateArray(double[] arr, int shift)
    {
        var result = new double[arr.Length];
        for (int i = 0; i < arr.Length; i++)
            result[i] = arr[(i + shift) % arr.Length];
        return result;
    }

    private double ComputeCorrelation(double[] a, double[] b)
    {
        double sumAB = 0, sumA2 = 0, sumB2 = 0;
        for (int i = 0; i < a.Length; i++)
        {
            sumAB += a[i] * b[i];
            sumA2 += a[i] * a[i];
            sumB2 += b[i] * b[i];
        }

        double denominator = Math.Sqrt(sumA2) * Math.Sqrt(sumB2);
        return denominator > 0 ? sumAB / denominator : 0;
    }
}

File: Source/Core/KeyDetector.cpp
#include "KeyDetector.h"
#include "AudioDataProvider.h"
#include <cmath>

namespace ToneAndBeats
{

KeyResult KeyDetector::detect(const std::vector<float>& monoSamples, int sampleRate)
{
    return detectFromSamples(monoSamples, sampleRate);
}

KeyResult KeyDetector::detect(const File& file)
{
    AudioDataProvider provider;
    auto [samples, rate] = provider.loadMono(file);
    return detectFromSamples(samples, rate);
}

KeyResult KeyDetector::detect(const String& filePath)
{
    return detect(juce::File(filePath));
}

KeyResult KeyDetector::detectFromSamples(const std::vector<float>& monoSamples, int sampleRate)
{
    KeyResult result = {"Unknown", "Unknown", "Unknown", "Unknown", 0.0};

    if (monoSamples.size() < static_cast<size_t>(sampleRate))
        return result;

    const int maxAnalysisSeconds = 30;
    int maxSamples = maxAnalysisSeconds * sampleRate;
    
    std::vector<float> analysisData;
    if (monoSamples.size() > static_cast<size_t>(maxSamples))
    {
        size_t startOffset = (monoSamples.size() - maxSamples) / 2;
        analysisData = std::vector<float>(monoSamples.begin() + startOffset,
                                         monoSamples.begin() + startOffset + maxSamples);
    }
    else
    {
        analysisData = monoSamples;
    }

    const int fftOrder = 14; // 16384 points
    const int fftSize = 1 << fftOrder;
    const int hopSize = fftSize / 2;
    const double a4Freq = 440.0;
    
    juce::dsp::FFT fft(fftOrder);
    juce::dsp::WindowingFunction<float> window(fftSize, juce::dsp::WindowingFunction<float>::hann);
    
    std::vector<double> pcp(12, 0.0);
    std::vector<float> fftData(fftSize * 2, 0.0f);
    int numFrames = 0;

    for (size_t frameStart = 0; frameStart + fftSize < analysisData.size(); frameStart += hopSize)
    {
        std::fill(fftData.begin(), fftData.end(), 0.0f);
        std::copy(analysisData.begin() + frameStart, analysisData.begin() + frameStart + fftSize, fftData.begin());
        window.multiplyWithWindowingTable(fftData.data(), fftSize);
        fft.performFrequencyOnlyForwardTransform(fftData.data());

        // According to report: Use magnitudes (sqrt of power)
        // Note: performFrequencyOnlyForwardTransform in JUCE gives magnitudes.

        for (int pitchClass = 0; pitchClass < 12; pitchClass++)
        {
            double pitchEnergy = 0;
            // Calculate base frequency for this note (relative to A440)
            double c0Freq = a4Freq * std::pow(2.0, (pitchClass - 9) / 12.0);
            
            // Octave wrapping: we want to check multiple octaves of this note
            // The report does it by checking 8 harmonics of the base note.
            for (int harmonic = 1; harmonic <= 8; harmonic++)
            {
                double harmonicFreq = c0Freq * harmonic;
                
                // Bring into a reasonable range if needed or just use as is
                // Actually, the C# code uses c0Freq as the base and checks harmonics.
                // But we should check all octaves of this pitch class in the musical range.
                // Let's follow the C# logic: harmonicFreq = c0Freq * harmonic
                
                while (harmonicFreq < 50.0) harmonicFreq *= 2.0; // Ensure it's in audible range
                if (harmonicFreq > (sampleRate / 2.0)) break;

                int freqBin = static_cast<int>(harmonicFreq * fftSize / sampleRate);
                if (freqBin > 0 && freqBin < (fftSize / 2) - 1)
                {
                    // Spectral Interpolation (mitigates quantization errors)
                    double interpolatedMag = static_cast<double>(fftData[freqBin]) + 
                                            0.5 * (static_cast<double>(fftData[freqBin - 1]) + static_cast<double>(fftData[freqBin + 1]));
                    
                    pitchEnergy += interpolatedMag / static_cast<double>(harmonic);
                }
            }
            pcp[pitchClass] += pitchEnergy;
        }
        numFrames++;
    }

    if (numFrames > 0)
    {
        double totalEnergy = 0;
        for (int i = 0; i < 12; i++)
        {
            pcp[i] /= numFrames;
            totalEnergy += pcp[i];
        }
        
        if (totalEnergy > 1e-6)
        {
            for (int i = 0; i < 12; i++)
                pcp[i] /= totalEnergy;
        }

        auto best = findNote(pcp);
        
        result.key = noteNames[best.root];
        result.mode = best.mode;
        result.confidence = std::min(1.0, best.score); // Correlation-based confidence

        // Calculate Relative
        if (best.mode == "Major")
        {
            int relIdx = (best.root + 9) % 12;
            result.relativeKey = noteNames[relIdx];
            result.relativeMode = "Minor";
        }
        else
        {
            int relIdx = (best.root + 3) % 12;
            result.relativeKey = noteNames[relIdx];
            result.relativeMode = "Major";
        }
    }

    return result;
}

KeyDetector::InternalKey KeyDetector::findNote(const std::vector<double>& pcp)
{
    double majorProfile[12] = { 6.35, 2.23, 3.48, 2.33, 4.38, 4.09, 2.52, 5.19, 2.39, 3.66, 2.29, 2.88 };
    double minorProfile[12] = { 6.33, 2.68, 3.52, 5.38, 2.60, 3.53, 2.54, 4.75, 3.98, 2.69, 3.34, 3.17 };
    
    InternalKey best = { 0, "Major", -1.0 };

    for (int root = 0; root < 12; root++)
    {
        // Rotate PCP to align 'root' with the profiles' 0th index
        std::vector<double> rotatedPcp(12);
        for (int i = 0; i < 12; i++)
            rotatedPcp[i] = pcp[(i + root) % 12];

        // Compare with Major
        double majorCorr = 0, sumA2 = 0, sumB2 = 0;
        for (int i = 0; i < 12; i++)
        {
            majorCorr += rotatedPcp[i] * majorProfile[i];
            sumA2 += rotatedPcp[i] * rotatedPcp[i];
            sumB2 += majorProfile[i] * majorProfile[i];
        }
        double majorSim = (sumA2 > 0 && sumB2 > 0) ? (majorCorr / (std::sqrt(sumA2) * std::sqrt(sumB2))) : 0;

        if (majorSim > best.score)
        {
            best.score = majorSim;
            best.root = root;
            best.mode = "Major";
        }

        // Compare with Minor
        double minorCorr = 0;
        sumB2 = 0; // sumA2 stays same
        for (int i = 0; i < 12; i++)
        {
            minorCorr += rotatedPcp[i] * minorProfile[i];
            sumB2 += minorProfile[i] * minorProfile[i];
        }
        double minorSim = (sumA2 > 0 && sumB2 > 0) ? (minorCorr / (std::sqrt(sumA2) * std::sqrt(sumB2))) : 0;

        if (minorSim > best.score)
        {
            best.score = minorSim;
            best.root = root;
            best.mode = "Minor";
        }
    }

    return best;
}

} // namespace ToneAndBeats